tests/utils/test.py

```diff
@@ -160,3 +167,81 @@ def create_fake_postcode():
     # Corsica and some DROM-COM department being 3 digits long,
     # return value may be longer than 5. Cut it.
     return f"{department}{int(random.randint(0, 999)):03}"[:5]
+
+
+class _AssertSnapshotQueriesContext(CaptureQueriesContext):
+    def __init__(self, snapshot, connection):
+        self.snapshot = snapshot
+        super().__init__(connection)
+
+    @staticmethod
+    def normalize_sql(sql):
+        if re.match(r'^SAVEPOINT +"[^"]+" *$', sql):
+            # 'SAVEPOINT "s124109847980928_x70"'
+            return 'SAVEPOINT "<savepoint>"'
+        if re.match(r'^RELEASE +SAVEPOINT +"[^"]+" *$', sql):
+            # 'RELEASE SAVEPOINT "s124109847980928_x69"'
+            return 'RELEASE SAVEPOINT "<savepoint>"'
+        return sql
+
+    def build_snapshot(self, queries):
+        return [{"sql": self.normalize_sql(query["raw_sql"]), "origin": query["origin"]} for query in queries]
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        super().__exit__(exc_type, exc_value, traceback)
+        if exc_type is not None:
+            return
+        assert self.build_snapshot(self.captured_queries) == self.snapshot
+
+
+def assertSnapshotQueries(snapshot, func=None, *args, using=DEFAULT_DB_ALIAS, **kwargs):
+    conn = connections[using]
+
+    context = _AssertSnapshotQueriesContext(snapshot, conn)
+    if func is None:
+        return context
+
+    with context:
+        func(*args, **kwargs)
+
+
+origin_debug_sql = CursorDebugWrapper.debug_sql
+
+
+def _detect_origin():
+    parts = []
+    # Ignore first 3 frames:
+    # - tests/utils/test.py::_detect_origin
+    # - tests/utils/test.py::debug_sql
+    # - <python>::contextlib.py::__exit__
+    for frame_info in inspect.stack()[3:]:
+        frame_filename = frame_info.filename
+        if frame_info.function == "render" and frame_filename.endswith("base.py"):
+            try:
+                origin = frame_info.frame.f_locals["self"].origin.template_name
+            except (AttributeError, KeyError):
+                origin = None
+            if origin is not None:
+                parts.append(f"Template[{origin}]")
+        elif frame_filename.startswith(settings.APPS_DIR):
+            filename = os.path.relpath(frame_filename, settings.APPS_DIR)
+            if "self" in frame_info.frame.f_locals:
+                class_name = type(frame_info.frame.f_locals["self"]).__name__
+                function = f"{class_name}.{frame_info.function}"
+            else:
+                function = frame_info.function
+            parts.append(f"{function}[{filename}]")
```

Comment:
Peut-être donner des précisions pour les longues fonctions ?
```suggestion
            parts.append(f"{function}[{filename}]:{frame_info.line_no}")
```

Reply:
J'ai peur que cela génère pas mal de diff sur les snapshots dès qu'on va rajouter une ligne dans le fichier...
Je pourrais éventuellement utiliser https://docs.python.org/3/library/linecache.html pour récupérer la ligne en elle-même, mais à voir comment on pourrait formater cela joliment :-/

Reply:
Ouais la `line_no` va être trop casse-pieds, dès qu'on va ajouter un import dans un fichier (ou déplacer/ajouter/supprimer du code) on va invalider des snapshots pour le code du fichier et donc avoir des tests qui cassent sur des trucs possiblement sans rapport.

Le code de la ligne va peut-être être un peu trop verbeux dans certain cas, et sur les trucs simple possible qu'on puisse en avoir plusieurs dans la même fonction, par exemple des `if ...` dans une vue.

Je pense que ça va pas être super pratique car c'est chiant à calculer mais vu qu'on affiche le nom de la fonction, on pourrais pas afficher le numéro de ligne du corps de la fonction ou tout autre truc relatif à elle ? C'est un peu un entre-deux quoi.